src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1318 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "common-thread-private.h"
+#include "mongoc-server-monitor-private.h"
+
+#include "mongoc/mongoc-client-private.h"
+#include "mongoc/mongoc-error-private.h"
+#include "mongoc/mongoc-interrupt-private.h"
+#include "mongoc/mongoc-ssl-private.h"
+#include "mongoc/mongoc-stream-private.h"
+#include "mongoc/mongoc-topology-background-monitoring-private.h"
+#include "mongoc/mongoc-topology-private.h"
+#include "mongoc/mongoc-trace-private.h"
+
+#undef MONGOC_LOG_DOMAIN
+#define MONGOC_LOG_DOMAIN "monitor"
+
+typedef enum {
+   MONGOC_THREAD_OFF = 0,
+   MONGOC_THREAD_RUNNING,
+   MONGOC_THREAD_SHUTTING_DOWN,
+   MONGOC_THREAD_JOINABLE
+} thread_state_t;
+
+/* Use a signed and wide return type for timeouts as long as you can. Cast only
+ * when you know what you're doing with it. */
+static int64_t
+_now_us (void)
+{
+   return bson_get_monotonic_time ();
+}
+
+static int64_t
+_now_ms (void)
+{
+   return _now_us () / 1000;
+}
+
+struct _mongoc_server_monitor_t {
+   mongoc_topology_t *topology;
+   bson_thread_t thread;
+
+   /* State accessed from multiple threads. */
+   struct {
+      bson_mutex_t mutex;
+      mongoc_cond_t cond;
+      thread_state_t state;
+      bool scan_requested;
+      bool cancel_requested;
+      mongoc_interrupt_t *interrupt;
+   } shared;
+
+   /* Default time to sleep between ismaster checks (reduced when a scan is
+    * requested) */
+   uint64_t heartbeat_frequency_ms;
```

Comment:
TODO: consider not copying everything from topology.

Reply:
Some tests may modify these on the shared topology, so this might not be so straightforward.

Reply:
min_heartbeat_frequency_msec is only overriden in tests before scanning starts. So copying the value on the creation of a server monitor should be ok. My preference is not to access the shared topology object directly outside of the topology mutex.